File: 654.最大二叉树.js
/*
 * @lc app=leetcode.cn id=654 lang=javascript
 *
 * [654] 最大二叉树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number[]} nums
 * @return {TreeNode}
 */
var constructMaximumBinaryTree = function (nums) {
  if (!nums.length) return null;

  const max = Math.max(...nums);
  let node = new TreeNode(max);
  if (nums.length === 1) return node;

  let deleteIndex;
  for (deleteIndex = 0; deleteIndex < nums.length; deleteIndex++)
    if (nums[deleteIndex] === max) break;

  const numsLeft = nums.slice(0, deleteIndex);
  const numsRight = nums.slice(deleteIndex + 1, nums.length);

  node.left = constructMaximumBinaryTree(numsLeft);
  node.right = constructMaximumBinaryTree(numsRight);

  return node;
};
// @lc code=end

File: store/654.最大二叉树.c
/*
 * @lc app=leetcode.cn id=654 lang=c
 *
 * [654] 最大二叉树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

struct TreeNode *back(int *nums, int start, int end)
{
  int index, max = -1, i;
  for (i = start; i <= end; i++)
  {
    if (max < nums[i])
    {
      max = nums[i];
      index = i;
    }
  }

  struct TreeNode *node = (struct TreeNode *)malloc(sizeof(struct TreeNode));
  node->left = NULL;
  node->right = NULL;
  node->val = max;

  if (start == end)
    return node;

  if (index - 1 >= start)
    node->left = back(nums, start, index - 1);

  if (index + 1 <= end)
    node->right = back(nums, index + 1, end);

  return node;
}

struct TreeNode *constructMaximumBinaryTree(int *nums, int numsSize)
{
  return back(nums, 0, numsSize - 1);
}
// @lc code=end
